stock_picking_mail_incorrect_qty/models/stock_picking.py

```diff
@@ -0,0 +1,130 @@
+# Copyright 2021 Coop IT Easy SCRL fs
+#   Carmen Bianca Bakker <carmen@coopiteasy.be>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, fields, models
+from odoo.tools.translate import _
+
+
+class Picking(models.Model):
+    _inherit = "stock.picking"
+
+    zero_received_move_line_ids = fields.One2many(
+        "stock.move.line",
+        string="Stock move lines that were not received",
+        compute="_compute_move_discrepancy",
+    )
+    zero_expected_move_line_ids = fields.One2many(
+        "stock.move.line",
+        string="Stock move lines of which some were received, but none were expected",
+        compute="_compute_move_discrepancy",
+    )
+    too_few_received_move_line_ids = fields.One2many(
+        "stock.move.line",
+        string="Stock move lines of which too few were received",
+        compute="_compute_move_discrepancy",
+    )
+    too_many_received_move_line_ids = fields.One2many(
+        "stock.move.line",
+        string="Stock move lines of which too many were received",
+        compute="_compute_move_discrepancy",
+    )
+
+    @api.multi
+    @api.depends("move_line_ids")
+    def _compute_move_discrepancy(self):
+        for picking in self:
+            # The following filter is also used in upstream action_done()
+            filtered_moves = picking.mapped("move_line_ids").filtered(
+                lambda line: line.state
+                in [
+                    "draft",
+                    "waiting",
+                    "partially_available",
+                    "assigned",
+                    "confirmed",
+                ]
+            )
+
+            picking.zero_received_move_line_ids = filtered_moves.filtered(
+                lambda line: line.qty_done == 0 and line.product_qty
+            )
+            picking.zero_expected_move_line_ids = filtered_moves.filtered(
+                lambda line: line.qty_done and line.product_qty == 0
+            )
+            picking.too_few_received_move_line_ids = filtered_moves.filtered(
+                lambda line: line.qty_done != 0
+                and line.qty_done < line.product_qty
+            )
+            picking.too_many_received_move_line_ids = filtered_moves.filtered(
+                lambda line: line.qty_done > line.product_qty
+                and line.product_qty != 0
+            )
+
+    @api.multi
+    def action_done(self):
+        # The following code between the 'fmt' tags is copied almost-verbatim
+        # from the stock module. It handles a corner case of move.lines being
+        # unlinked to moves, but linked to stock.picking. It is copied here
+        # verbatim because not including it would (potentially) result in some
+        # moves being skipped over.
```

Comment:
The order I want:

- Copied code snippet from stock_picking.action_done(). This fixes discrepancies as described.
- Send e-mail
- The remainder of stock_picking.action_done(). This function 'destroys' the information that is needed for composing the e-mail (e.g., empty moves are cancelled, quantities are split and adjusted, etc etc etc).

If I do the following order:

- Send e-mail
- All of stock_picking.action_done()

… then the e-mailing functionality may not have access to the fixed discrepancies.

_However_, because I refactored this to use stock.move.lines instead of stock.moves\*, there may be a chance that we are now no longer affected by the possible discrepancies. So I guess this can be removed?

\*: I don't particularly like this change. The module worked fine without the change … just not on bees.test.

Reply:
> The order I want:
>  ...

Ok I get it. Since it's in addons, I guess we can commit this. Your explanation about order of execution would be welcome in the comments as well.

> _However_, because I refactored this to use stock.move.lines instead of stock.moves*, there may be a chance that we are now no longer affected by the possible discrepancies. So I guess this can be removed?

Not sure about that, I see _rabbit holes_ everywhere in stocks now, we'll need to test.

> *: I don't particularly like this change. The module worked fine without the change … just not on bees.test.

Did you check in the shell the modules used on bees.test ? I remember of a module to invoice of product from picking lines that were not in the sale order. I cannot find the name of the module right now and scrum is calling.

Thanks for your investigations

Reply:
@robinkeunen I added the above explanation as a comment to the code, and I have reverted the change from stock.move to stock.move.line. Turns out I was using the wrong variable for what I really wanted. Why it displayed different behaviour on bees compared to not-bees, I may never know. I couldn't find any module with `odoo_repl` that altered any relevant code, and I crawled up the code tree quite a bit.